lib/api.ml

```diff
@@ -52,36 +60,47 @@ type 'a api_request =
       (request_method
       * endpoint
       * request_args
+      * maybe_body
       * 'a Decoder.Yojson.Safe.decoder)
 
-let make_api_request request_method endpoint args request_decoder =
-  Api_request (request_method, endpoint, args, request_decoder)
+let make_api_request request_method endpoint args maybe_body request_decoder =
+  Api_request (request_method, endpoint, args, maybe_body, request_decoder)
 ;;
 
 let make_api_get_request ?(args = []) endpoint request_decoder =
-  make_api_request GET endpoint args request_decoder
+  make_api_request GET endpoint args None request_decoder
+;;
+
+let make_api_post_request ?(args = []) endpoint body request_decoder =
+  make_api_request POST endpoint args (Some body) request_decoder
 ;;
 
 let ( --> ) m json_body_fn =
   let ( >>= ) = Lwt.bind in
   let ( >|= ) = Lwt.Infix.( >|= ) in
   m
-  >>= fun (_, body) ->
-  Cohttp_lwt.Body.to_string body
-  >|= fun body -> Yojson.Safe.from_string body |> json_body_fn
+  >>= fun (response, body) ->
+  let status_code = Cohttp.(Response.status response |> Code.code_of_status) in
+  let string_body = Cohttp_lwt.Body.to_string body in
+  if Cohttp.Code.is_success status_code then
+    (string_body >|= fun body -> Yojson.Safe.from_string body |> json_body_fn)
+  else
+    (string_body >|= fun body -> Decoder.Yojson.Safe.fail (Printf.sprintf "Status Code %d: %s" status_code body) (Yojson.Safe.from_string body))
```

Comment:
I think it's a good idea to harden this and deal with other response codes. However, I don't think that this should result in a JSON decoder error -- the actual failure lies somewhere else. Maybe the result should be of type result where the error part differentiates between the different kinds of errors (e.g. connection error, json decoder error, ...), something like:

```ocaml
type response_error =
  | Http_error of int
  | Json_decoder_error of Decoder.Yojson.Safe.Error.t
  | ...
```

Reply:
Addressed this with [22fb949](https://github.com/neshtea/kimai_report/pull/1/commits/22fb9498f8d91c87ebb729795a23683d8c9726c1)